### PR TITLE
Add `let_underscore_must_use` lint

### LINKED ISSUE #4812: New lint: let _ binding of must use function
Extracted from #4655

A lint that detects `let _` bindings of functions, that are `must_use`:
```rust
#[must_use]
fn foo() -> u32 { 42 }

fn bar() -> Result<(), ()> { Ok(()) }

let _ = foo();
let _ = bar();
```

On how to detect must use https://github.com/rust-lang/rust-clippy/issues/4655#issuecomment-544104531:
>We have a check for `#[must_use]` types in functions.rs already (IIRC `is_must_use_ty(cx, ty)`, we best move this to `utils/mod.rs`), that you can call with the result of an `expr_ty(cx, _)` on the initializer.

_Originally posted by @flip1995 in https://github.com/rust-lang/rust-clippy/issues/4655#issuecomment-541024037_

### PR DESCRIPTION
**Pull request by Areredify**
changelog: closes #4812 , added a new `let_underscore_must_use` lint, moved `is_must_use_ty` to utils, added `is_must_use_fn` util function

### CODE REVIEW

**Comment by Areredify** [comment]
ping @flip1995 , is this fail my mistake or not?

**Comment by llogiq** [comment]
Toolstate says clippy master is currently broken, so it's probably not your mistake.

**Comment by RalfJung** [comment]
Yeah this is https://github.com/laumann/compiletest-rs/issues/202

File: clippy_lints/src/let_underscore.rs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    /// fn f() -> Result<u32, u32> {
+    ///     Ok(0)
+    /// }
+    ///
+    /// let _ = f();
+    /// // is_ok() is marked #[must_use]
+    /// let _ = f().is_ok();
+    /// ```
+    pub LET_UNDERSCORE_MUST_USE,
+    pedantic,
```
Comment (flip1995) [inline clippy_lints/src/let_underscore.rs]:
```suggestion
    restriction,
```

This was requested as a restriction lint, and I agree that it should be a restriction lint.

File: tests/ui/let_underscore.rs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+impl Trait for S {
+    fn a() -> u32 {
+        0
+    }
+}
+
+fn main() {
+    let _ = f();
```
Comment (flip1995) [inline tests/ui/let_underscore.rs]: Please add a test for a function, that is not `must_use` and doesn't have a `must_use` type.

**Comment by flip1995** [comment]
Thanks

@bors r+

**Comment by flip1995** [comment]
`./util/dev update_lints` required

**Comment by flip1995** [comment]
Sorry for my absence. Can you rebase this, so we can merge this?

**Comment by Areredify** [comment]
Done

**Pull request merged**